### ISSUE TITLE
[Messenger] Allow to provide context in the Serializer class

### ISSUE THREAD

**Issue by romain-pierre** [symfony/symfony#26900]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | no
| Symfony version  | 4.1

In the class `Symfony\Component\Messenger\Transport\Serialization`, the method `encode($message): array` serializes the message with an empty context option. It might be interesting to be able to provide some context options (from the configuration for example) instead of overloading all the class.

**Comment by chalasr** [comment]
Making it Implement ContextAwareEncoder|DecoderInterface looks sensible to me.
\cc @sroze

**Comment by sroze** [comment]
Yes, that's a good idea. Though, just two parameters (for the format and the serialization context) in [the declaration of the service](https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml#L46-L48) would be enough, nah? Alternatively we could have the parameter configured via the `framework.messenger` configuration.

**Comment by romain-pierre** [comment]
I don't see any restriction to configure a context in the configuration (so, under `framework.messenger`) but it's just not usual. Or only allow some keys and do the context construction in the class?

As an example:

```yaml
framework:
    messenger:
        transport:
            serialize_message_with_type_in_headers:
                enable_max_depth: true
                groups: [messenger]
                disable_type_enforcement: ~
```

Or yes, just add a setter on the Serialize service, to inject a custom context. So we can configure in the service definition, or set (and change) at runtime, in a middleware for example (custom logic, adaptive to a message).

**Comment by sroze** [comment]
I was just thinking about a parameter... so you'd configure it this way:
```yaml
parameters:
    messenger.transport.default_serialization_context: {your: 'options'}
```

**Comment by ogizanagi** [comment]
Would make sense to add a messenger config option for this indeed.
AFAIK, DI parameters aren't an extension point we support (appart for a few cases) and aren't good for discoverability anyway.

Another way (or complementary) would be to make it configurable at runtime by wrapping the original message into a context aware one. The same could be useful for some specific use-cases where middlewares might need to be configured at runtime for a specific message, e.g validation groups.

**Comment by sroze** [comment]
Based on the fact that this transport is the default one and very likely to be the one used in 90%+ of the scenarios, I'd say it makes sense to be able to configure its options via the `framework.messenger` configuration. I'll issue a PR ASAP for that one.

**Comment by sroze** [comment]
Fixed by #26941
